dot/parachain/spam_slots.go

```diff
@@ -0,0 +1,200 @@
+	return slotKey{
+		session:   session,
+		validator: validator,
+	}
+}
+
+func unconfirmedDisputesKey(session parachain.SessionIndex, candidate common.Hash) unconfirmedKey {
+	return unconfirmedKey{
+		session:   session,
+		candidate: candidate,
+	}
+}
+
+func (s *spamSlots) getSpamCount(session parachain.SessionIndex, validator parachain.ValidatorIndex) (uint32, bool) {
+	s.slotsLock.RLock()
+	defer s.slotsLock.RUnlock()
+	spamCount, ok := s.slots[slotsKey(session, validator)]
+	return spamCount, ok
+}
+
+func (s *spamSlots) getValidators(session parachain.SessionIndex, candidate common.Hash) (*treeset.Set, bool) {
+	s.unconfirmedLock.RLock()
+	defer s.unconfirmedLock.RUnlock()
+	validators, ok := s.unconfirmed[unconfirmedDisputesKey(session, candidate)]
+	return validators, ok
+}
+
+func (s *spamSlots) clearValidators(session parachain.SessionIndex, candidate common.Hash) {
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+	delete(s.unconfirmed, unconfirmedDisputesKey(session, candidate))
+}
+
+func (s *spamSlots) clearSlots(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	delete(s.slots, slotsKey(session, validator))
+}
+
+func (s *spamSlots) incrementSpamCount(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	s.slots[slotsKey(session, validator)]++
+}
+
+func (s *spamSlots) AddUnconfirmed(session parachain.SessionIndex, candidate common.Hash, validator parachain.ValidatorIndex) bool {
+	if spamCount, _ := s.getSpamCount(session, validator); spamCount >= s.maxSpamVotes {
+		return false
+	}
+
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+
+	validators, ok := s.unconfirmed[unconfirmedDisputesKey(session, candidate)]
+	if !ok || validators == nil {
+		validators = treeset.NewWith(byValidatorIndex)
+	}
+
+	if !validators.Contains(validator) {
+		validators.Add(validator)
+		s.incrementSpamCount(session, validator)
+	}
+
+	s.unconfirmed[unconfirmedDisputesKey(session, candidate)] = validators
+
+	return true
+}
+
+func (s *spamSlots) Clear(session parachain.SessionIndex, candidate common.Hash) {
+	if validators, ok := s.getValidators(session, candidate); ok {
+		validatorSet := validators.Values()
+		s.clearValidators(session, candidate)
+
+		for _, validator := range validatorSet {
+			spamCount, ok := s.getSpamCount(session, validator.(parachain.ValidatorIndex))
+			if ok {
+				if spamCount == 1 {
+					s.clearSlots(session, validator.(parachain.ValidatorIndex))
+					continue
+				}
+
+				s.slotsLock.Lock()
+				s.slots[slotsKey(session, validator.(parachain.ValidatorIndex))] = spamCount - 1
+				s.slotsLock.Unlock()
+			}
+		}
+	}
+}
+
+func (s *spamSlots) PruneOld(oldestIndex parachain.SessionIndex) {
+	s.unconfirmedLock.Lock()
+	unconfirmedToDelete := make([]unconfirmedKey, 0)
```

Comment:
Since the max items `unconfirmedToDelete` can append is all items in the `s.unconfirmed` I would to suggest to pass it as the capacity of the slice
```suggestion
	unconfirmedToDelete := make([]unconfirmedKey, 0, len(s.unconfirmed))
```

Reply:
I've changed this to use `maps.DeleteFunc` instead.